src/renderer/renderScene.ts

```diff
@@ -328,468 +328,541 @@ export const _renderScene = ({
+  if (renderGrid && appState.gridSize) {
+    strokeGrid(
+      context,
+      appState.gridSize,
+      -Math.ceil(renderConfig.zoom.value / appState.gridSize) *
+        appState.gridSize +
+        (renderConfig.scrollX % appState.gridSize),
+      -Math.ceil(renderConfig.zoom.value / appState.gridSize) *
+        appState.gridSize +
+        (renderConfig.scrollY % appState.gridSize),
+      normalizedCanvasWidth / renderConfig.zoom.value,
+      normalizedCanvasHeight / renderConfig.zoom.value,
     );
+  }
 
-    let editingLinearElement: NonDeleted<ExcalidrawLinearElement> | undefined =
-      undefined;
-    visibleElements.forEach((element) => {
-      try {
-        renderElement(element, rc, context, renderConfig, appState);
-        // Getting the element using LinearElementEditor during collab mismatches version - being one head of visible elements due to
-        // ShapeCache returns empty hence making sure that we get the
-        // correct element from visible elements
-        if (appState.editingLinearElement?.elementId === element.id) {
-          if (element) {
-            editingLinearElement =
-              element as NonDeleted<ExcalidrawLinearElement>;
-          }
-        }
-        if (!isExporting) {
-          renderLinkIcon(element, context, appState);
+  // Paint visible elements
+  const visibleElements = elements.filter((element) =>
+    isVisibleElement(element, normalizedCanvasWidth, normalizedCanvasHeight, {
+      zoom: renderConfig.zoom,
+      offsetLeft: appState.offsetLeft,
+      offsetTop: appState.offsetTop,
+      scrollX: renderConfig.scrollX,
+      scrollY: renderConfig.scrollY,
+    }),
+  );
+
+  let editingLinearElement: NonDeleted<ExcalidrawLinearElement> | undefined =
+    undefined;
+  visibleElements.forEach((element) => {
+    try {
+      renderElement(element, rc, context, renderConfig, appState);
+      // Getting the element using LinearElementEditor during collab mismatches version - being one head of visible elements due to
+      // ShapeCache returns empty hence making sure that we get the
+      // correct element from visible elements
+      if (appState.editingLinearElement?.elementId === element.id) {
+        if (element) {
+          editingLinearElement = element as NonDeleted<ExcalidrawLinearElement>;
         }
-      } catch (error: any) {
-        console.error(error);
       }
-    });
+      if (!isExporting) {
+        renderLinkIcon(element, context, appState);
+      }
+    } catch (error: any) {
+      console.error(error);
+    }
+  });
 
-    if (editingLinearElement) {
-      renderLinearPointHandles(
+  if (editingLinearElement) {
+    renderLinearPointHandles(
+      context,
+      appState,
+      renderConfig,
+      editingLinearElement,
+    );
+  }
+  context.restore();
+  context.restore();
+
+  return { atLeastOneVisibleElement: visibleElements.length > 0 };
+};
+
+export const renderCanvasUI = ({
+  elements,
+  appState,
+  scale,
+  rc,
```

Comment:
let's remove the `rc` dependency (and thus factor out the selection element rendering)

Reply:
`rc` is is not of use under `renderCanvasUI`, but both are using `renderElement` which uses `rc`

Reply:
> `rc` is is not of use under `renderCanvasUI`, but both are using `renderElement` which uses `rc`

yep, hence:

> [...] (and thus factor out the selection element rendering)

as it's not using `rc`

Reply:
this one is resolved, just verify it once, have removed `rc` from `renderCanvasUI`